File: leetcodes/121. Best Time to Buy and Sell Stock/maxProfit.java
class Solution {
  public int maxProfit(int[] prices) {
    //input: array
    //output: integer
    /**
     * one pass:
     * iterate the array,
     *    get the min value,
     *    if not min, get the max(max, current - minvalue)
     * time O(n) space O(1)
     */
    int max = 0;
    int min = prices[0];
    for (int i = 1; i < prices.length; i++) {
      if (min > prices[i]) {
        min = prices[i];
      } else {
        max = Math.max(max, prices[i] - min);
      }
    }
    return max;
  }
}

File: leetcodes/121. Best Time to Buy and Sell Stock/maxProfit.ts
function maxProfit(prices: number[]): number {
  //time O(n) space O(1)
  let maxN: number = 0
  let minN = Number.MAX_VALUE
  let prof: number = 0
  for (let i= 0; i < prices.length; i++ ){
    if (prices[i] < minN){
      minN = prices[i]
      maxN = prices[i]
    } else if (prices[i] > maxN) {
      maxN = prices[i]
    }
    prof = Math.max(maxN - minN, prof)
  }
  return prof
};